File: macos-helper/FlameConnectTokenHelper.m
#import <AppKit/AppKit.h>
#import <CommonCrypto/CommonDigest.h>
#import <CoreServices/CoreServices.h>
#import <Security/Security.h>

static NSString * const ClientID = @"1af761dc-085a-411f-9cb9-53e5e2115bd2";
static NSString * const CallbackScheme = @"msal1af761dc-085a-411f-9cb9-53e5e2115bd2";
static NSString * const RedirectURI = @"msal1af761dc-085a-411f-9cb9-53e5e2115bd2://auth";
static NSString * const Authority = @"https://gdhvb2cflameconnect.b2clogin.com/gdhvb2cflameconnect.onmicrosoft.com/B2C_1A_FirePhoneSignUpOrSignInWithPhoneOrEmail";
static NSString * const Scope = @"openid profile offline_access https://gdhvb2cflameconnect.onmicrosoft.com/Mobile/read";
static NSString * const HelperBundleID = @"com.dibsies.FlameConnectTokenHelper";

static NSString *Base64URL(NSData *data) {
    NSString *value = [data base64EncodedStringWithOptions:0];
    value = [value stringByReplacingOccurrencesOfString:@"+" withString:@"-"];
    value = [value stringByReplacingOccurrencesOfString:@"/" withString:@"_"];
    return [value stringByReplacingOccurrencesOfString:@"=" withString:@""];
}

static NSString *RandomURLSafe(NSUInteger count) {
    NSMutableData *data = [NSMutableData dataWithLength:count];
    if (SecRandomCopyBytes(kSecRandomDefault, count, data.mutableBytes) != errSecSuccess) abort();
    return Base64URL(data);
}

static NSString *SHA256URLSafe(NSString *value) {
    NSData *input = [value dataUsingEncoding:NSUTF8StringEncoding];
    unsigned char digest[CC_SHA256_DIGEST_LENGTH];
    CC_SHA256(input.bytes, (CC_LONG)input.length, digest);
    return Base64URL([NSData dataWithBytes:digest length:sizeof(digest)]);
}

static NSDictionary<NSString *, NSString *> *QueryValues(NSURL *url) {
    NSMutableDictionary *values = [NSMutableDictionary dictionary];
    for (NSURLQueryItem *item in [NSURLComponents componentsWithURL:url resolvingAgainstBaseURL:NO].queryItems ?: @[]) {
        values[item.name] = item.value ?: @"";
    }
    return values;
}

@interface AppDelegate : NSObject <NSApplicationDelegate>
@property NSWindow *window;
@property NSTextField *statusLabel;
@property NSButton *tokenButton;
@property NSString *state;
@property NSString *verifier;
@property NSString *refreshToken;
@property NSString *previousHandler;
@end

@implementation AppDelegate

- (void)applicationDidFinishLaunching:(NSNotification *)notification {
    [self buildWindow];
    if ([[NSProcessInfo processInfo].arguments containsObject:@"--self-test"]) {
        [self runSelfTest];
        return;
    }
    [self.window makeKeyAndOrderFront:nil];
    [NSApp activateIgnoringOtherApps:YES];
    [self startAuthorization];
}

- (void)application:(NSApplication *)application openURLs:(NSArray<NSURL *> *)urls {
    for (NSURL *url in urls) {
        if ([url.scheme isEqualToString:CallbackScheme]) {
            [self handleCallback:url];
            break;
        }
    }
}

- (void)applicationWillTerminate:(NSNotification *)notification { [self restoreHandler]; }

- (void)buildWindow {
    self.window = [[NSWindow alloc] initWithContentRect:NSMakeRect(0, 0, 540, 240)
                                              styleMask:NSWindowStyleMaskTitled | NSWindowStyleMaskClosable | NSWindowStyleMaskMiniaturizable
                                                backing:NSBackingStoreBuffered defer:NO];
    self.window.title = @"Flame Connect Token Helper";
    [self.window center];

    NSTextField *title = [NSTextField labelWithString:@"Flame Connect sign-in"];
    title.font = [NSFont boldSystemFontOfSize:22];
    title.alignment = NSTextAlignmentCenter;
    self.statusLabel = [NSTextField wrappingLabelWithString:@"Preparing secure sign-in…"];
    self.statusLabel.alignment = NSTextAlignmentCenter;
    self.statusLabel.maximumNumberOfLines = 4;

    NSButton *start = [NSButton buttonWithTitle:@"Start Again" target:self action:@selector(startAuthorization)];
    start.bezelStyle = NSBezelStyleRounded;
    self.tokenButton = [NSButton buttonWithTitle:@"Copy Refresh Token" target:self action:@selector(copyToken)];
    self.tokenButton.bezelStyle = NSBezelStyleRounded;
    self.tokenButton.hidden = YES;

    NSStackView *buttons = [NSStackView stackViewWithViews:@[start, self.tokenButton]];
    buttons.orientation = NSUserInterfaceLayoutOrientationHorizontal;
    buttons.spacing = 12;
    NSStackView *stack = [NSStackView stackViewWithViews:@[title, self.statusLabel, buttons]];
    stack.orientation = NSUserInterfaceLayoutOrientationVertical;
    stack.spacing = 22;
    stack.alignment = NSLayoutAttributeCenterX;
    stack.translatesAutoresizingMaskIntoConstraints = NO;
    [self.window.contentView addSubview:stack];
    [NSLayoutConstraint activateConstraints:@[
        [stack.leadingAnchor constraintEqualToAnchor:self.window.contentView.leadingAnchor constant:32],
        [stack.trailingAnchor constraintEqualToAnchor:self.window.contentView.trailingAnchor constant:-32],
        [stack.centerYAnchor constraintEqualToAnchor:self.window.contentView.centerYAnchor],
    ]];
}

- (NSURL *)authorizationURL {
    self.verifier = RandomURLSafe(48);
    self.state = RandomURLSafe(24);
    NSString *nonce = RandomURLSafe(24);
    NSURLComponents *components = [NSURLComponents componentsWithString:[Authority stringByAppendingString:@"/oauth2/v2.0/authorize"]];
    components.queryItems = @[
        [NSURLQueryItem queryItemWithName:@"client_id" value:ClientID],
        [NSURLQueryItem queryItemWithName:@"client_info" value:@"1"],
        [NSURLQueryItem queryItemWithName:@"response_type" value:@"code"],
        [NSURLQueryItem queryItemWithName:@"redirect_uri" value:RedirectURI],
        [NSURLQueryItem queryItemWithName:@"response_mode" value:@"query"],
        [NSURLQueryItem queryItemWithName:@"scope" value:Scope],
        [NSURLQueryItem queryItemWithName:@"state" value:self.state],
        [NSURLQueryItem queryItemWithName:@"nonce" value:SHA256URLSafe(nonce)],
        [NSURLQueryItem queryItemWithName:@"code_challenge" value:SHA256URLSafe(self.verifier)],
        [NSURLQueryItem queryItemWithName:@"code_challenge_method" value:@"S256"],
    ];
    return components.URL;
}

- (void)claimHandler {
    if (!self.previousHandler) {
        CFStringRef current = LSCopyDefaultHandlerForURLScheme((__bridge CFStringRef)CallbackScheme);
        if (current) {
            NSString *handler = CFBridgingRelease(current);
            if (![handler isEqualToString:HelperBundleID]) self.previousHandler = handler;
        }
    }
    LSSetDefaultHandlerForURLScheme((__bridge CFStringRef)CallbackScheme, (__bridge CFStringRef)HelperBundleID);
}

- (void)restoreHandler {
    if (self.previousHandler) {
        LSSetDefaultHandlerForURLScheme((__bridge CFStringRef)CallbackScheme, (__bridge CFStringRef)self.previousHandler);
        self.previousHandler = nil;
    }
}

- (void)startAuthorization {
    self.refreshToken = nil;
    self.tokenButton.hidden = YES;
    [self claimHandler];
    self.statusLabel.stringValue = @"Safari will open. Sign in, then choose Allow when asked to open this helper.";
    [[NSWorkspace sharedWorkspace] openURL:[self authorizationURL]];
}

- (void)handleCallback:(NSURL *)url {
    NSDictionary *values = QueryValues(url);
    NSString *error = values[@"error_description"] ?: values[@"error"];
    if (error.length) { [self showError:error]; return; }
    if (![values[@"state"] isEqualToString:self.state]) { [self showError:@"The security state did not match. Please start again."]; return; }
    NSString *code = values[@"code"];
    if (!code.length) { [self showError:@"The callback did not contain an authorization code."]; return; }
    self.statusLabel.stringValue = @"Finishing sign-in…";
    [self exchangeCode:code];
}

- (void)exchangeCode:(NSString *)code {
    NSURL *url = [NSURL URLWithString:[Authority stringByAppendingString:@"/oauth2/v2.0/token"]];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    request.HTTPMethod = @"POST";
    [request setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Content-Type"];
    NSURLComponents *body = [[NSURLComponents alloc] init];
    body.queryItems = @[
        [NSURLQueryItem queryItemWithName:@"client_id" value:ClientID],
        [NSURLQueryItem queryItemWithName:@"client_info" value:@"1"],
        [NSURLQueryItem queryItemWithName:@"grant_type" value:@"authorization_code"],
        [NSURLQueryItem queryItemWithName:@"code" value:code],
        [NSURLQueryItem queryItemWithName:@"redirect_uri" value:RedirectURI],
        [NSURLQueryItem queryItemWithName:@"code_verifier" value:self.verifier],
        [NSURLQueryItem queryItemWithName:@"scope" value:Scope],
    ];
    request.HTTPBody = [body.percentEncodedQuery dataUsingEncoding:NSUTF8StringEncoding];
    [[[NSURLSession sharedSession] dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *networkError) {
        dispatch_async(dispatch_get_main_queue(), ^{
            if (networkError) { [self showError:networkError.localizedDescription]; return; }
            NSDictionary *json = data ? [NSJSONSerialization JSONObjectWithData:data options:0 error:nil] : nil;
            NSString *error = json[@"error_description"] ?: json[@"error"];
            if (error.length) { [self showError:error]; return; }
            NSString *token = json[@"refresh_token"];
            if (!token.length) { [self showError:@"Sign-in succeeded but no refresh token was returned."]; return; }
            self.refreshToken = token;
            [self restoreHandler];
            self.tokenButton.hidden = NO;
            self.statusLabel.stringValue = @"Success. Copy the refresh token, then paste it into Homebridge → Flame Connect → Refresh Token.";
            [NSApp activateIgnoringOtherApps:YES];
            [self.window makeKeyAndOrderFront:nil];
        });
    }] resume];
}

- (void)copyToken {
    if (!self.refreshToken.length) return;
    [[NSPasteboard generalPasteboard] clearContents];
    [[NSPasteboard generalPasteboard] setString:self.refreshToken forType:NSPasteboardTypeString];
    self.statusLabel.stringValue = @"Refresh token copied. Paste it into Homebridge → Flame Connect → Refresh Token.";
}

- (void)showError:(NSString *)message {
    self.statusLabel.stringValue = message ?: @"Unknown error";
    self.tokenButton.hidden = YES;
    [NSApp activateIgnoringOtherApps:YES];
    [self.window makeKeyAndOrderFront:nil];
}

- (void)runSelfTest {
    NSURL *url = [self authorizationURL];
    NSDictionary *values = QueryValues(url);
    BOOL passed = [values[@"client_info"] isEqualToString:@"1"] &&
        [values[@"state"] isEqualToString:self.state] &&
        [values[@"redirect_uri"] isEqualToString:RedirectURI] &&
        [values[@"nonce"] length] > 0 && [values[@"code_challenge"] length] > 0;
    fprintf(stdout, "%s\n", passed ? "Self-test passed" : "Self-test failed");
    [NSApp terminate:nil];
    if (!passed) exit(1);
}
@end

int main(int argc, const char *argv[]) {
    @autoreleasepool {
        NSApplication *application = [NSApplication sharedApplication];
        AppDelegate *delegate = [[AppDelegate alloc] init];
        application.delegate = delegate;
        [application setActivationPolicy:NSApplicationActivationPolicyRegular];
        [application run];
    }
    return 0;
}

File: macos-helper/FlameConnectTokenHelper.swift
import AppKit
import CryptoKit
import CoreServices
import Foundation

private let clientID = "1af761dc-085a-411f-9cb9-53e5e2115bd2"
private let callbackScheme = "msal\(clientID)"
private let redirectURI = "\(callbackScheme)://auth"
private let authority = "https://gdhvb2cflameconnect.b2clogin.com/gdhvb2cflameconnect.onmicrosoft.com/B2C_1A_FirePhoneSignUpOrSignInWithPhoneOrEmail"
private let scope = "openid profile offline_access https://gdhvb2cflameconnect.onmicrosoft.com/Mobile/read"
private let helperBundleID = "com.dibsies.FlameConnectTokenHelper"

private func base64URL(_ data: Data) -> String {
    data.base64EncodedString()
        .replacingOccurrences(of: "+", with: "-")
        .replacingOccurrences(of: "/", with: "_")
        .replacingOccurrences(of: "=", with: "")
}

private func randomURLSafe(byteCount: Int) -> String {
    var bytes = [UInt8](repeating: 0, count: byteCount)
    guard SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes) == errSecSuccess else {
        fatalError("Secure random-number generation failed")
    }
    return base64URL(Data(bytes))
}

private func sha256URLSafe(_ value: String) -> String {
    base64URL(Data(SHA256.hash(data: Data(value.utf8))))
}

private struct AuthFlow {
    let state: String
    let verifier: String
    let authorizationURL: URL

    static func create() throws -> AuthFlow {
        let verifier = randomURLSafe(byteCount: 48)
        let state = randomURLSafe(byteCount: 24)
        let nonce = randomURLSafe(byteCount: 24)
        var components = URLComponents(string: "\(authority)/oauth2/v2.0/authorize")!
        components.queryItems = [
            URLQueryItem(name: "client_id", value: clientID),
            URLQueryItem(name: "client_info", value: "1"),
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "redirect_uri", value: redirectURI),
            URLQueryItem(name: "response_mode", value: "query"),
            URLQueryItem(name: "scope", value: scope),
            URLQueryItem(name: "state", value: state),
            URLQueryItem(name: "nonce", value: sha256URLSafe(nonce)),
            URLQueryItem(name: "code_challenge", value: sha256URLSafe(verifier)),
            URLQueryItem(name: "code_challenge_method", value: "S256"),
        ]
        guard let url = components.url else { throw HelperError.invalidAuthorizationURL }
        return AuthFlow(state: state, verifier: verifier, authorizationURL: url)
    }
}

private enum HelperError: LocalizedError {
    case invalidAuthorizationURL
    case invalidCallback
    case stateMismatch
    case authorization(String)
    case token(String)
    case missingRefreshToken

    var errorDescription: String? {
        switch self {
        case .invalidAuthorizationURL: return "Could not create the Flame Connect authorization URL."
        case .invalidCallback: return "The Flame Connect callback did not contain an authorization code."
        case .stateMismatch: return "The security state did not match. Please start the login again."
        case .authorization(let message): return message
        case .token(let message): return "Token exchange failed: \(message)"
        case .missingRefreshToken: return "Flame Connect signed in, but did not return a refresh token."
        }
    }
}

@MainActor
final class AppDelegate: NSObject, NSApplicationDelegate {
    private var window: NSWindow!
    private var statusLabel: NSTextField!
    private var startButton: NSButton!
    private var copyButton: NSButton!
    private var flow: AuthFlow?
    private var refreshToken: String?
    private var previousHandler: String?

    func applicationDidFinishLaunching(_ notification: Notification) {
        buildWindow()
        if CommandLine.arguments.contains("--self-test") {
            runSelfTestAndExit()
            return
        }
        window.makeKeyAndOrderFront(nil)
        NSApp.activate(ignoringOtherApps: true)
        startAuthorization()
    }

    func application(_ application: NSApplication, open urls: [URL]) {
        guard let url = urls.first(where: { $0.scheme == callbackScheme }) else { return }
        handleCallback(url)
    }

    func applicationWillTerminate(_ notification: Notification) {
        restorePreviousHandler()
    }

    private func buildWindow() {
        window = NSWindow(
            contentRect: NSRect(x: 0, y: 0, width: 520, height: 230),
            styleMask: [.titled, .closable, .miniaturizable],
            backing: .buffered,
            defer: false
        )
        window.title = "Flame Connect Token Helper"
        window.center()

        let title = NSTextField(labelWithString: "Flame Connect sign-in")
        title.font = .boldSystemFont(ofSize: 22)
        title.alignment = .center

        statusLabel = NSTextField(wrappingLabelWithString: "Preparing secure sign-in…")
        statusLabel.alignment = .center
        statusLabel.maximumNumberOfLines = 4

        startButton = NSButton(title: "Start Again", target: self, action: #selector(startAgain))
        startButton.bezelStyle = .rounded

        copyButton = NSButton(title: "Copy Refresh Token", target: self, action: #selector(copyToken))
        copyButton.bezelStyle = .rounded
        copyButton.isHidden = true

        let buttons = NSStackView(views: [startButton, copyButton])
        buttons.orientation = .horizontal
        buttons.spacing = 12
        buttons.alignment = .centerY

        let stack = NSStackView(views: [title, statusLabel, buttons])
        stack.orientation = .vertical
        stack.spacing = 22
        stack.alignment = .centerX
        stack.translatesAutoresizingMaskIntoConstraints = false
        window.contentView?.addSubview(stack)
        NSLayoutConstraint.activate([
            stack.leadingAnchor.constraint(equalTo: window.contentView!.leadingAnchor, constant: 32),
            stack.trailingAnchor.constraint(equalTo: window.contentView!.trailingAnchor, constant: -32),
            stack.centerYAnchor.constraint(equalTo: window.contentView!.centerYAnchor),
        ])
    }

    @objc private func startAgain() { startAuthorization() }

    @objc private func copyToken() {
        guard let refreshToken else { return }
        NSPasteboard.general.clearContents()
        NSPasteboard.general.setString(refreshToken, forType: .string)
        statusLabel.stringValue = "Refresh token copied. Paste it into Homebridge → Flame Connect → Refresh Token."
    }

    private func startAuthorization() {
        do {
            refreshToken = nil
            copyButton.isHidden = true
            flow = try AuthFlow.create()
            claimCallbackScheme()
            statusLabel.stringValue = "Safari will open. Sign in, then choose Allow when asked to open this helper."
            NSWorkspace.shared.open(flow!.authorizationURL)
        } catch {
            show(error)
        }
    }

    private func claimCallbackScheme() {
        if previousHandler == nil,
           let current = LSCopyDefaultHandlerForURLScheme(callbackScheme as CFString)?.takeRetainedValue() as String?,
           current != helperBundleID {
            previousHandler = current
        }
        LSSetDefaultHandlerForURLScheme(callbackScheme as CFString, helperBundleID as CFString)
    }

    private func restorePreviousHandler() {
        if let previousHandler {
            LSSetDefaultHandlerForURLScheme(callbackScheme as CFString, previousHandler as CFString)
            self.previousHandler = nil
        }
    }

    private func handleCallback(_ url: URL) {
        guard let flow else { return }
        do {
            let items = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems ?? []
            let values = Dictionary(uniqueKeysWithValues: items.map { ($0.name, $0.value ?? "") })
            if let description = values["error_description"], !description.isEmpty {
                throw HelperError.authorization(description)
            }
            if let error = values["error"], !error.isEmpty { throw HelperError.authorization(error) }
            guard values["state"] == flow.state else { throw HelperError.stateMismatch }
            guard let code = values["code"], !code.isEmpty else { throw HelperError.invalidCallback }
            statusLabel.stringValue = "Finishing sign-in…"
            exchange(code: code, verifier: flow.verifier)
        } catch {
            show(error)
        }
    }

    private func exchange(code: String, verifier: String) {
        let endpoint = URL(string: "\(authority)/oauth2/v2.0/token")!
        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        let fields = [
            "client_id": clientID,
            "client_info": "1",
            "grant_type": "authorization_code",
            "code": code,
            "redirect_uri": redirectURI,
            "code_verifier": verifier,
            "scope": scope,
        ]
        var components = URLComponents()
        components.queryItems = fields.map { URLQueryItem(name: $0.key, value: $0.value) }
        request.httpBody = components.percentEncodedQuery?.data(using: .utf8)

        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            Task { @MainActor in
                guard let self else { return }
                do {
                    if let error { throw HelperError.token(error.localizedDescription) }
                    guard let data else { throw HelperError.token("No response received.") }
                    let object = try JSONSerialization.jsonObject(with: data) as? [String: Any]
                    if let description = object?["error_description"] as? String {
                        throw HelperError.token(description)
                    }
                    guard let token = object?["refresh_token"] as? String, !token.isEmpty else {
                        throw HelperError.missingRefreshToken
                    }
                    self.refreshToken = token
                    self.restorePreviousHandler()
                    self.copyButton.isHidden = false
                    self.statusLabel.stringValue = "Success. Copy the refresh token, then paste it into the Homebridge plugin settings."
                    NSApp.activate(ignoringOtherApps: true)
                    self.window.makeKeyAndOrderFront(nil)
                } catch {
                    self.show(error)
                }
            }
        }.resume()
    }

    private func show(_ error: Error) {
        statusLabel.stringValue = error.localizedDescription
        copyButton.isHidden = true
        NSApp.activate(ignoringOtherApps: true)
        window.makeKeyAndOrderFront(nil)
    }

    private func runSelfTestAndExit() {
        do {
            let testFlow = try AuthFlow.create()
            let components = URLComponents(url: testFlow.authorizationURL, resolvingAgainstBaseURL: false)
            let values = Dictionary(uniqueKeysWithValues: (components?.queryItems ?? []).map { ($0.name, $0.value ?? "") })
            precondition(values["client_info"] == "1")
            precondition(values["state"] == testFlow.state)
            precondition(values["nonce"]?.isEmpty == false)
            precondition(values["code_challenge"]?.isEmpty == false)
            precondition(values["redirect_uri"] == redirectURI)
            print("Self-test passed")
            NSApp.terminate(nil)
        } catch {
            fputs("Self-test failed: \(error.localizedDescription)\n", stderr)
            exit(1)
        }
    }
}

let application = NSApplication.shared
let delegate = AppDelegate()
application.delegate = delegate
application.setActivationPolicy(.regular)
application.run()
